rasterio/_base.pyx

```diff
@@ -439,6 +440,7 @@ cdef class DatasetBase:
         self._closed = True
 
     def __enter__(self):
+        self._env.enter_context(env_ctx_if_needed())
```

Comment:
I feel dumb for not seeing this until now. I could have done this for 1.3.0!

Reply:
`dataset.enter_context(Env())` is like a retroactive version of 

```
with Env():
    with dataset as dataset:
        ...
```

It lets us add an outer context (`Env()`) to the inner one (`dataset`).